### PR TITLE
PATE Differential Privacy

### PR DESCRIPTION
**Pull request by iamtrask**
This adds the PATE DP analysis to our codebase with a unit test showing it's functionality

### CODE REVIEW

File: syft/frameworks/torch/differential_privacy/pate.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
+# Copyright 2016 The TensorFlow Authors. All Rights Reserved.
```
Comment (robert-wagner) [inline syft/frameworks/torch/differential_privacy/pate.py]: Does this copyright belong here?

File: syft/frameworks/torch/differential_privacy/pate.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+reproduce key results from our paper (https://arxiv.org/abs/1610.05755):
+the epsilon bounds for MNIST and SVHN students.
+
+The command that computes the epsilon bound associated
+with the training of the MNIST student model (100 label queries
+with a (1/20)*2=0.1 epsilon bound each) is:
+
+python analysis.py
+  --counts_file=mnist_250_teachers_labels.npy
+  --indices_file=mnist_250_teachers_100_indices_used_by_student.npy
```
Comment (robert-wagner) [inline syft/frameworks/torch/differential_privacy/pate.py]: This command does not look like it should be here

File: syft/frameworks/torch/differential_privacy/pate.py [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+
+python analysis.py
+  --counts_file=svhn_250_teachers_labels.npy
+  --max_examples=1000
+  --delta=1e-6
+"""
+import os
+import math
+import numpy as np
+from six.moves import xrange
```
Comment (robert-wagner) [inline syft/frameworks/torch/differential_privacy/pate.py]: Do we need six since this is a python 3 only project

File: test/torch/federated/test_dataset.py [hunk 4]
```diff
@@ -27,6 +25,9 @@ def test_base_dataset(workers):
         pass
 
 
+from syft.frameworks.torch.federated import BaseDataset
```
Comment (robert-wagner) [inline test/torch/federated/test_dataset.py]: Why was this import moved

File: syft/frameworks/torch/differential_privacy/pate.py (hunk 1 above)
Comment (iamtrask) [reply syft/frameworks/torch/differential_privacy/pate.py]: I *think* so - when you include just one file written (primarily) by other authors I think you just throw the copyright on the top of the file.

File: syft/frameworks/torch/differential_privacy/pate.py (hunk 2 above)
Comment (iamtrask) [reply syft/frameworks/torch/differential_privacy/pate.py]: so true

Comment (iamtrask) [reply syft/frameworks/torch/differential_privacy/pate.py]: fixed

File: syft/frameworks/torch/differential_privacy/pate.py (hunk 3 above)
Comment (iamtrask) [reply syft/frameworks/torch/differential_privacy/pate.py]: fixed

File: test/torch/federated/test_dataset.py (hunk 4 above)
Comment (iamtrask) [reply test/torch/federated/test_dataset.py]: accident lol

Comment (iamtrask) [reply test/torch/federated/test_dataset.py]: i think i just fat fingered some keys

Comment (iamtrask) [reply test/torch/federated/test_dataset.py]: fixed

File: examples/experimental/PATE Analysis.ipynb [hunk 5]
```diff
@@ -0,0 +1,84 @@
+{
```
Comment (LaRiffle) [inline examples/experimental/PATE Analysis.ipynb]: This notebook is empty, did you want to add an example of the code you produced, or didn't want to add this nb?

File: examples/experimental/PointersAhead.ipynb [hunk 6]
```diff
@@ -0,0 +1,435 @@
+{
```
Comment (LaRiffle) [inline examples/experimental/PointersAhead.ipynb]: Maybe remove this notebook from the PR as it is unrelated

File: examples/experimental/PATE Analysis.ipynb (hunk 5 above)
Comment (iamtrask) [reply examples/experimental/PATE Analysis.ipynb]: deleted the notebook

File: examples/experimental/PointersAhead.ipynb (hunk 6 above)
Comment (iamtrask) [reply examples/experimental/PointersAhead.ipynb]: deleted the notebook
